### PR TITLE
Scala support

### PR DESCRIPTION
**Pull request by mikekap**
Adds barebones scala_library & scala_test (still junit) support. This is sort of experimental - Scala compiles a bit differently than java (i.e. dependencies can be executed as part of the compile), so I'm not sure this works in all cases. It does work for the simple cases though.

LMK if you'd like to document this support or leave it until it's a bit more solid.

### CODE REVIEW

**Comment by sdwilsh** [comment]
I'm fine with not documenting this until we (well, you) feel it is more solid.

**Comment by mikekap** [comment]
Particularly, zinc is almost a necessity with any size scala project. It works a bit differently than the normal compiler. It's an incremental compiler to the extent that it wants the previously compiled class files & uses a nailgun daemon itself.

I'm not sure how to integrate that into buck yet (other than "externally"), but certainly some scala support is better than no scala support :)

File: src/com/facebook/buck/jvm/scala/ScalaBuckConfig.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+  public ScalaBuckConfig(final BuckConfig delegate) {
+    this.delegate = delegate;
+  }
+
+  public Tool getScalac(BuildRuleResolver resolver) {
+    CommandTool.Builder scalac = new CommandTool.Builder(findScalac(resolver));
+
+    // Add some standard options.
+    scalac.addArg("-target:" + delegate.getValue("scala", "target_level").or("jvm-1.7"));
```
Comment (ruibm) [inline src/com/facebook/buck/jvm/scala/ScalaBuckConfig.java]: Can you make "scala" a constant please?

File: src/com/facebook/buck/jvm/scala/ScalaBuckConfig.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    String scalaHome = delegate.getEnvironment().get("SCALA_HOME");
+    if (scalaHome != null) {
+      Path scalacInHomePath = Paths.get(scalaHome, "bin", "scalac");
+      if (scalacInHomePath.toFile().exists()) {
+        return new HashedFileTool(scalacInHomePath);
+      }
+    }
+
+    throw new HumanReadableException("Could not find scalac. Consider setting scala.compiler.");
```
Comment (ruibm) [inline src/com/facebook/buck/jvm/scala/ScalaBuckConfig.java]: Could you please have two different messages for the case where SCALA_HOME is not defined and the case when it's defined in the env however the file does not exist.

File: src/com/facebook/buck/jvm/scala/ScalaLibraryDescription.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    // Note: scala does not have a exported_deps because scala needs the transitive closure of
+    // dependencies to compile. deps is effectively exported_deps.
+    public Optional<ImmutableSortedSet<BuildTarget>> providedDeps;
+    public Optional<ImmutableSortedSet<BuildTarget>> deps;
+
+    @Hint(isInput = false)
+    public Optional<Path> resourcesRoot;
+    public Optional<String> mavenCoords;
+
+    @Hint(isDep = false) public Optional<ImmutableSortedSet<BuildTarget>> tests;
```
Comment (ruibm) [inline src/com/facebook/buck/jvm/scala/ScalaLibraryDescription.java]: For consistency, either keep the @Hint annotation in the same line or breakline like in line #163.

**Comment by ruibm** [comment]
Thanks for sending this PR. Looks good - I just added few minor comments.

**Comment by ruibm** [comment]
Looks good! Thanks for the changes. Proceeding with the merge.

**Pull request closed without merging**